### PR TITLE
add inputs_embeds to Bart/MBart/Unified_Transformer/Unimo/CodeGen

### PR DESCRIPTION
**Pull request by Yam0214**
### PR types

New features
### PR changes

Models
### Description

1. add `inputs_embeds` to Bart/MBart/Unified_Transformer/Unimo/CodeGen
2. force `use_cache` to `False` if `labels` is provided for Bart and MBart Model (in training)

### CODE REVIEW

**Comment by Yam0214** [comment]
@wj-Mcat

File: paddlenlp/transformers/bart/modeling.py [hunk 1]
```diff
@@ -515,11 +540,13 @@ def set_input_embeddings(self, value):
 
     def forward(
         self,
-        input_ids,
+        input_ids=None,
```
Comment (FrostML) [inline paddlenlp/transformers/bart/modeling.py:543]: 也修改下文档，加一个 `optional`，后面的地方也一样。

File: paddlenlp/transformers/mbart/modeling.py [hunk 2]
```diff
@@ -645,13 +709,19 @@ def forward(
...
+                raise ValueError(
+                    "If no `decoder_input_ids` or `decoder_inputs_embeds` are "
+                    "passed, `input_ids` cannot be `None`. Please pass either "
+                    "`input_ids` or `decoder_input_ids` or `decoder_inputs_embeds`."
+                )
             decoder_input_ids = shift_tokens_right(input_ids, self.pad_token_id)
-        if attention_mask is None:
-            assert input_ids is not None, "input_ids should be " "specified when generating attention_mask"
+        if attention_mask is None and input_ids is not None:
+            # assert input_ids is not None, "input_ids should be " \
```
Comment (FrostML) [inline paddlenlp/transformers/mbart/modeling.py]: 这里能否也和 codegen 的修改一样，在 input_ids 为 None 的时候贴一个 warning，然后删除这个注释

File: paddlenlp/transformers/unified_transformer/modeling.py [hunk 3]
```diff
@@ -169,32 +169,52 @@ def __init__(
...
         self.pad_token_id = pad_token_id
 
-    def forward(self, input_ids, token_type_ids=None, position_ids=None, role_ids=None):
+    def forward(self, input_ids, token_type_ids=None, position_ids=None, role_ids=None, input_embeddings=None):
+        if input_ids is None and input_embeddings is None:
+            raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
+        elif input_ids is not None:
+            inputs_sample = input_ids
+        elif input_embeddings is not None:
+            inputs_sample = input_embeddings[:, :, -1]
```
Comment (FrostML) [inline paddlenlp/transformers/unified_transformer/modeling.py]: 这里的 `inputs_sample` 在其表示意义上有点奇怪，后面的 `paddle.expand_as()` 能否替换成 `paddle.expand()`，然后这里改成获取 `input_shape`，也避免后续反复调用 `paddle.shape()`

File: paddlenlp/transformers/unified_transformer/modeling.py [hunk 4]
```diff
@@ -169,32 +169,52 @@ def __init__(
...
+                    num_pad = paddle.sum((input_ids == self.pad_token_id).astype("float32"), axis=-1, keepdim=True)
+                    position_ids = F.relu(
+                        paddle.expand_as(
+                            paddle.arange(end=paddle.shape(inputs_sample)[1], dtype="float32"), inputs_sample
+                        )
+                        - num_pad
+                    ).astype("int64")
+                else:
+                    logger.warning(
+                        "position_ids or pad_token_ids should be provided when input_embeds is specified, otherwise an unexpected result may be returned"
```
Comment (FrostML) [inline paddlenlp/transformers/unified_transformer/modeling.py]: otherwise 这句，说明下是怎样的 position id 吧，`an unexpected result` 有些笼统了

File: paddlenlp/transformers/codegen/modeling.py [hunk 5]
```diff
@@ -488,30 +500,34 @@ def forward(
...
+                    attention_mask = (
+                        paddle.cast(input_ids == self.pad_token_id, dtype=paddle.get_default_dtype()).unsqueeze([1, 2])
+                        * -1e4
+                    )
             else:
-                attention_mask = (
-                    paddle.cast(input_ids == self.pad_token_id, dtype=paddle.get_default_dtype()).unsqueeze([1, 2])
-                    * -1e4
-                )
+                logger.warning("provided inputs_embeds without attention_mask")
```
Comment (FrostML) [inline paddlenlp/transformers/codegen/modeling.py]: 首字母大写，其次说明下将会使用默认值 None 作为 attention mask，表示不进行 mask 操作。后面的也一样。

File: paddlenlp/transformers/unimo/modeling.py [hunk 6]
```diff
@@ -242,20 +242,39 @@ def __init__(
...
 
-    def forward(self, input_ids, token_type_ids=None, position_ids=None):
-        input_embedings = self.word_embeddings(input_ids)
+    def forward(self, input_ids=None, token_type_ids=None, position_ids=None, input_embeddings=None):
+        if input_ids is None and input_embeddings is None:
+            raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
+        elif input_ids is not None:
+            inputs_sample = input_ids
+        elif input_embeddings is not None:
+            inputs_sample = input_embeddings[:, :, -1]
```
Comment (FrostML) [inline paddlenlp/transformers/unimo/modeling.py]: 这里和上面一样，需要修改下。

Comment (Yam0214) [reply paddlenlp/transformers/unimo/modeling.py]: 1. input_ids 的文档加上 `optional`
2. `assert input_ids is not None` 注释删除，并加上logger
3. `paddle.expand_as` 部分替换成 `paddle.expand`
4. 补充了 `logger.warning`的内容，开头大写。

**Review by FrostML — APPROVED** [review]
